versioned_docs/version-2.0/tutorials/rag.mdx

```diff
@@ -0,0 +1,545 @@
+    """
+
+    # Get specific steps in our RAG pipeline, which are noted with @traceable decorator
+    rag_pipeline_run = next(
+        run for run in root_run.child_runs if run.name == "get_answer"
+    )
+    retrieve_run = next(
+        run for run in rag_pipeline_run.child_runs if run.name == "retrieve_docs"
+    )
+    contexts = "\n\n".join(doc.page_content for doc in retrieve_run.outputs["output"])
+    input_question = example.inputs["input_question"]
+
+    # LLM grader
+    llm = ChatOpenAI(model="gpt-4-turbo", temperature=0)
+
+    # Structured prompt
+    answer_grader = grade_prompt_doc_relevance | llm
+
+    # Get score
+    score = answer_grader.invoke({"question":input_question,
+                                  "documents":contexts})
+    score = score["Score"]
+
+    return {"key": "document_relevance", "score": score}
+
+def answer_hallucination_grader(root_run: Run, example: Example) -> dict:
+    """
+    A simple evaluator that checks to see the answer is grounded in the documents
+    """
+
+    # RAG input
+    rag_pipeline_run = next(
+        run for run in root_run.child_runs if run.name == "get_answer"
+    )
+    retrieve_run = next(
+        run for run in rag_pipeline_run.child_runs if run.name == "retrieve_docs"
+    )
+    contexts = "\n\n".join(doc.page_content for doc in retrieve_run.outputs["output"])
+
+    # RAG output
+    prediction = rag_pipeline_run.outputs["answer"]
+
+    # LLM grader
+    llm = ChatOpenAI(model="gpt-4-turbo", temperature=0)
+
+    # Structured prompt
+    answer_grader = grade_prompt_hallucinations | llm
+
+    # Get score
+    score = answer_grader.invoke({"documents": contexts,
+                                  "student_answer": prediction})
+    score = score["Score"]
+
+    return {"key": "answer_hallucination", "score": score}
+
+experiment_results = evaluate(
+    predict_rag_answer,
+    data=dataset_name,
+    evaluators=[document_relevance_grader, answer_hallucination_grader],
+    metadata={"version": "LCEL context, gpt-4-0125-preview"},
+)
+```
+
+    View the evaluation results for experiment: 'passionate-stomach-89' at:
+    https://smith.langchain.com/o/ebbaf2eb-769b-4505-aca2-d11de10372a4/datasets/5f34de91-78ab-4ae2-83f3-8f7f93afc998/compare?selectedSessions=72d9c039-84ff-468a-a537-e9a33f8b7e92
+
+
+
+
+
+    0it [00:00, ?it/s]
+
+## Online Evaluation
+
+Three of our evaluators do not require a `reference` (ground truth) answer:
+
+- **Response vs input**
+- **Response vs retrieved docs**
+- **Retrieved docs vs input**
+
+These simply perform a comparison between run input (question), intermediate steps (documents), and outputs (answer).
+
+This allows us to use them as online evaluators.
+
+```python
+# Project
+os.environ["LANGCHAIN_PROJECT"] = "RAG_online_evaluation"
+```
+
+```python
+result = rag_bot.get_answer("How do I create a RAG chain?")
+```
+
+```python
+result = rag_bot.get_answer("How can I use a runnable lambda?")
+```
```

Comment:
To me this section doesn't really make sense - what does the code mean? I think it's good enough to just remove the code, as the text already seems to make sense without it?